### PR TITLE
Rebase of #90: Add Secure Boot support

### LINKED ISSUE #66: proxmox-clone does not honor os argument
When filing a bug, please include the following headings if possible. Any
example text in this template can be deleted.

#### Overview of the Issue

proxmox-clone does not honor `os` argument.

A paragraph or two about the issue you're experiencing.

#### Reproduction Steps

- Setup your VM template with a specific value for os type.
- Build a manifest with proxmox-clone invoking the template `clone_vm = "your-template"`, and `os` optional argument different than the source template one.
- You will see the source Os Type attribute not your settings one during building process. also in the final vm template. 

### Plugin and Packer version

Packer v1.7.10
packer-plugin-proxmox v1.0.4

### Simplified Packer Buildfile

<details>

```
packer {
  required_plugins {
    proxmox = {
      version = ">= 1.0.4"
      source  = "github.com/hashicorp/proxmox"
    }
  }
}

variable "proxmox_token" {
  type    = string
  default = "${env("PROXMOX_TOKEN")}"
}

source "proxmox-clone" "autogenerated_1" {
  clone_vm     = "ubuntu-2004-cloudimg"
  ssh_username = "ubuntu"
  insecure_skip_tls_verify = true
  
  ssh_host             = "192.168.0.144"
  network_adapters {
    bridge = "vmbr0"
    model  = "virtio"
    mac_address = "be:ef:ca:fe:c0:fe"
  }

  node                 = "pvelab"
  proxmox_url          = "https://foo.example.net:8006/api2/json"
  template_description = "image made from cloud-init image"
  template_name        = "ubuntu-2004-scaffolding"
  token                = "${var.proxmox_token}"
  username             = "root@pam!packer-test"
  memory               = 2048
  os                   = "l26"
  scsi_controller      = "virtio-scsi-pci"
}

build {
  sources = ["source.proxmox-clone.autogenerated_1"]
  provisioner "shell" {
    inline = ["echo test", "sleep 10", "echo bye"]
  }
}
```
</details>

### Operating system and Environment details

```
$ lsb_release -d
Description:	Ubuntu 20.04.4 LTS
$ uname -m
x86_64
```

### Log Fragments and crash.log files

This are the logs from packer. From proxmox side I saw the HTTP requests and I can confirm there is no setup with `ostype` attribute for the VM nor the Template.

<details>

```
2022/02/22 19:33:08 [INFO] Packer version: 1.7.10 [go1.17.6 linux amd64]
2022/02/22 19:33:08 [TRACE] discovering plugins in /home/linuxbrew/.linuxbrew/Cellar/packer/1.7.10/libexec/bin
2022/02/22 19:33:08 [TRACE] discovering plugins in /home/user/.config/packer/plugins
2022/02/22 19:33:08 [TRACE] discovering plugins in .
2022/02/22 19:33:08 [DEBUG] Discovered plugin: custom = /home/user/Doing/2022_packer_proxmox/sandbox_packer_proxmox/packer-plugin-custom
2022/02/22 19:33:08 [INFO] found external [-packer-default-plugin-name- proxmox-clone proxmox-iso] builders from custom plugin
2022/02/22 19:33:08 [TRACE] discovering plugins in 
2022/02/22 19:33:08 [DEBUG] Discovered plugin: custom = /home/user/Doing/2022_packer_proxmox/sandbox_packer_proxmox/packer-plugin-custom
2022/02/22 19:33:08 [INFO] found external [-packer-default-plugin-name- proxmox-clone proxmox-iso] builders from custom plugin
2022/02/22 19:33:08 [TRACE] discovering plugins in /home/linuxbrew/.linuxbrew/bin
2022/02/22 19:33:08 [INFO] PACKER_CONFIG env var not set; checking the default config file path
2022/02/22 19:33:08 [INFO] PACKER_CONFIG env var set; attempting to open config file: /home/user/.packerconfig
2022/02/22 19:33:08 [WARN] Config file doesn't exist: /home/user/.packerconfig
2022/02/22 19:33:08 [INFO] Setting cache directory: /home/user/.cache/packer
2022/02/22 19:33:08 [TRACE] listing potential installations for "github.com/hashicorp/proxmox" that match ">= 1.0.4". plugingetter.ListInstallationsOptions{FromFolders:[]string{"/home/linuxbrew/.linuxbrew/Cellar/packer/1.7.10/libexec/bin/packer", ".", "/home/user/.config/packer/plugins", "", "/home/linuxbrew/.linuxbrew/bin"}, BinaryInstallationOptions:plugingetter.BinaryInstallationOptions{APIVersionMajor:"5", APIVersionMinor:"0", OS:"linux", ARCH:"amd64", Ext:"", Checksummers:[]plugingetter.Checksummer{plugingetter.Checksummer{Type:"sha256", Hash:(*sha256.digest)(0xc00018a100)}}}}
2022/02/22 19:33:08 [TRACE] Found the following "github.com/hashicorp/proxmox" installations: [{/home/linuxbrew/.linuxbrew/bin/github.com/hashicorp/proxmox/packer-plugin-proxmox_v1.0.4_x5.0_linux_amd64 v1.0.4}]
2022/02/22 19:33:08 [INFO] found external [-packer-default-plugin-name- proxmox-clone proxmox-iso] builders from proxmox plugin
2022/02/22 19:33:08 [TRACE] validateValue: not active for proxmox_token, so skipping
2022/02/22 19:33:08 [TRACE] validateValue: not active for proxmox_url, so skipping
2022/02/22 19:33:08 [TRACE] validateValue: not active for proxmox_username, so skipping
2022/02/22 19:33:08 [TRACE] validateValue: not active for proxmox_node, so skipping
2022/02/22 19:33:08 [TRACE] Starting internal plugin packer-builder-proxmox-clone
2022/02/22 19:33:08 Starting plugin: /home/linuxbrew/.linuxbrew/Cellar/packer/1.7.10/libexec/bin/packer []string{"/home/linuxbrew/.linuxbrew/Cellar/packer/1.7.10/libexec/bin/packer", "plugin", "packer-builder-proxmox-clone"}
2022/02/22 19:33:08 Waiting for RPC address for: /home/linuxbrew/.linuxbrew/Cellar/packer/1.7.10/libexec/bin/packer
2022/02/22 19:33:08 packer-builder-proxmox-clone plugin: [INFO] Packer version: 1.7.10 [go1.17.6 linux amd64]
2022/02/22 19:33:08 packer-builder-proxmox-clone plugin: [INFO] PACKER_CONFIG env var not set; checking the default config file path
2022/02/22 19:33:08 packer-builder-proxmox-clone plugin: [INFO] PACKER_CONFIG env var set; attempting to open config file: /home/user/.packerconfig
2022/02/22 19:33:08 packer-builder-proxmox-clone plugin: [WARN] Config file doesn't exist: /home/user/.packerconfig
2022/02/22 19:33:08 packer-builder-proxmox-clone plugin: [INFO] Setting cache directory: /home/user/.cache/packer
2022/02/22 19:33:08 packer-builder-proxmox-clone plugin: args: []string{"packer-builder-proxmox-clone"}
2022/02/22 19:33:08 packer-builder-proxmox-clone plugin: Plugin address: unix /tmp/packer-plugin4088226102
2022/02/22 19:33:08 packer-builder-proxmox-clone plugin: Waiting for connection...
2022/02/22 19:33:08 Received unix RPC address for /home/linuxbrew/.linuxbrew/Cellar/packer/1.7.10/libexec/bin/packer: addr is /tmp/packer-plugin4088226102
2022/02/22 19:33:08 packer-builder-proxmox-clone plugin: Serving a plugin connection...
2022/02/22 19:33:08 packer-builder-proxmox-clone plugin: Number of cores 0 is too small, using default: 1
2022/02/22 19:33:08 packer-builder-proxmox-clone plugin: Number of sockets 0 is too small, using default: 1
2022/02/22 19:33:08 packer-builder-proxmox-clone plugin: CPU type not set, using default 'kvm64'
2022/02/22 19:33:08 [TRACE] Starting internal plugin packer-provisioner-shell
2022/02/22 19:33:08 Starting plugin: /home/linuxbrew/.linuxbrew/Cellar/packer/1.7.10/libexec/bin/packer []string{"/home/linuxbrew/.linuxbrew/Cellar/packer/1.7.10/libexec/bin/packer", "plugin", "packer-provisioner-shell"}
2022/02/22 19:33:08 Waiting for RPC address for: /home/linuxbrew/.linuxbrew/Cellar/packer/1.7.10/libexec/bin/packer
2022/02/22 19:33:08 packer-provisioner-shell plugin: [INFO] Packer version: 1.7.10 [go1.17.6 linux amd64]
2022/02/22 19:33:08 packer-provisioner-shell plugin: [INFO] PACKER_CONFIG env var not set; checking the default config file path
2022/02/22 19:33:08 packer-provisioner-shell plugin: [INFO] PACKER_CONFIG env var set; attempting to open config file: /home/user/.packerconfig
2022/02/22 19:33:08 packer-provisioner-shell plugin: [WARN] Config file doesn't exist: /home/user/.packerconfig
2022/02/22 19:33:08 packer-provisioner-shell plugin: [INFO] Setting cache directory: /home/user/.cache/packer
2022/02/22 19:33:08 packer-provisioner-shell plugin: args: []string{"packer-provisioner-shell"}
2022/02/22 19:33:08 Received unix RPC address for /home/linuxbrew/.linuxbrew/Cellar/packer/1.7.10/libexec/bin/packer: addr is /tmp/packer-plugin3016092351
2022/02/22 19:33:08 packer-provisioner-shell plugin: Plugin address: unix /tmp/packer-plugin3016092351
2022/02/22 19:33:08 packer-provisioner-shell plugin: Waiting for connection...
2022/02/22 19:33:08 packer-provisioner-shell plugin: Serving a plugin connection...
2022/02/22 19:33:08 Build debug mode: false
2022/02/22 19:33:08 Force build: false
2022/02/22 19:33:08 On error: 
proxmox-clone.autogenerated_1: output will be in this color.
2022/02/22 19:33:08 Waiting on builds to complete...

2022/02/22 19:33:08 Starting build run: proxmox-clone.autogenerated_1
2022/02/22 19:33:08 Running builder: 
2022/02/22 19:33:08 [INFO] (telemetry) Starting builder proxmox-clone.autogenerated_1
2022/02/22 19:33:08 packer-builder-proxmox-clone plugin: using token auth
==> proxmox-clone.autogenerated_1: Creating ephemeral key pair for SSH communicator...
==> proxmox-clone.autogenerated_1: Created ephemeral SSH key pair for communicator
==> proxmox-clone.autogenerated_1: Creating VM
==> proxmox-clone.autogenerated_1: No VM ID given, getting next free from Proxmox
==> proxmox-clone.autogenerated_1: Starting VM
2022/02/22 19:33:21 packer-builder-proxmox-clone plugin: No boot command given, skipping
==> proxmox-clone.autogenerated_1: Using SSH communicator to connect: 192.168.0.144
2022/02/22 19:33:21 packer-builder-proxmox-clone plugin: [INFO] Waiting for SSH, up to timeout: 5m0s
==> proxmox-clone.autogenerated_1: Waiting for SSH to become available...
2022/02/22 19:33:36 packer-builder-proxmox-clone plugin: [DEBUG] TCP connection to SSH ip/port failed: dial tcp 192.168.0.144:22: i/o timeout
2022/02/22 19:33:41 packer-builder-proxmox-clone plugin: [INFO] Attempting SSH connection to 192.168.0.144:22...
2022/02/22 19:33:41 packer-builder-proxmox-clone plugin: [DEBUG] reconnecting to TCP connection for SSH
2022/02/22 19:33:41 packer-builder-proxmox-clone plugin: [DEBUG] handshaking with SSH
2022/02/22 19:33:42 packer-builder-proxmox-clone plugin: [DEBUG] handshake complete!
2022/02/22 19:33:42 packer-builder-proxmox-clone plugin: [DEBUG] Opening new ssh session
2022/02/22 19:33:44 packer-builder-proxmox-clone plugin: [INFO] agent forwarding enabled
==> proxmox-clone.autogenerated_1: Connected to SSH!
2022/02/22 19:33:44 packer-builder-proxmox-clone plugin: Running the provision hook
2022/02/22 19:33:44 [INFO] (telemetry) Starting provisioner shell
==> proxmox-clone.autogenerated_1: Provisioning with shell script: /tmp/packer-shell3898576573
2022/02/22 19:33:44 packer-provisioner-shell plugin: Opening /tmp/packer-shell3898576573 for reading
2022/02/22 19:33:44 packer-provisioner-shell plugin: [INFO] 41 bytes written for 'uploadData'
2022/02/22 19:33:44 [INFO] 41 bytes written for 'uploadData'
2022/02/22 19:33:44 packer-builder-proxmox-clone plugin: [DEBUG] Opening new ssh session
2022/02/22 19:33:44 packer-builder-proxmox-clone plugin: [DEBUG] Starting remote scp process:  scp -vt /tmp
2022/02/22 19:33:44 packer-builder-proxmox-clone plugin: [DEBUG] Started SCP session, beginning transfers...
2022/02/22 19:33:44 packer-builder-proxmox-clone plugin: [DEBUG] Copying input data into temporary file so we can read the length
2022/02/22 19:33:44 packer-builder-proxmox-clone plugin: [DEBUG] scp: Uploading script_9232.sh: perms=C0644 size=41
2022/02/22 19:33:44 packer-builder-proxmox-clone plugin: [DEBUG] SCP session complete, closing stdin pipe.
2022/02/22 19:33:44 packer-builder-proxmox-clone plugin: [DEBUG] Waiting for SSH session to complete.
2022/02/22 19:33:44 packer-builder-proxmox-clone plugin: [DEBUG] scp stderr (length 30): Sink: C0644 41 script_9232.sh
2022/02/22 19:33:44 packer-builder-proxmox-clone plugin: [DEBUG] Opening new ssh session
2022/02/22 19:33:44 packer-builder-proxmox-clone plugin: [DEBUG] starting remote command: chmod 0755 /tmp/script_9232.sh
2022/02/22 19:33:44 packer-builder-proxmox-clone plugin: [INFO] RPC endpoint: Communicator ended with: 0
2022/02/22 19:33:44 [INFO] RPC client: Communicator ended with: 0
2022/02/22 19:33:44 [INFO] RPC endpoint: Communicator ended with: 0
2022/02/22 19:33:44 packer-provisioner-shell plugin: [INFO] RPC client: Communicator ended with: 0
2022/02/22 19:33:44 packer-builder-proxmox-clone plugin: [DEBUG] Opening new ssh session
2022/02/22 19:33:44 packer-builder-proxmox-clone plugin: [DEBUG] starting remote command: chmod +x /tmp/script_9232.sh; PACKER_BUILDER_TYPE='proxmox-clone' PACKER_BUILD_NAME='autogenerated_1' PACKER_HTTP_PORT='0'  /tmp/script_9232.sh
    proxmox-clone.autogenerated_1: test
2022/02/22 19:33:54 packer-builder-proxmox-clone plugin: [INFO] RPC endpoint: Communicator ended with: 0
2022/02/22 19:33:54 [INFO] 0 bytes written for 'stderr'
2022/02/22 19:33:54 [INFO] 9 bytes written for 'stdout'
2022/02/22 19:33:54 [INFO] RPC client: Communicator ended with: 0
2022/02/22 19:33:54 [INFO] RPC endpoint: Communicator ended with: 0
2022/02/22 19:33:54 packer-provisioner-shell plugin: [INFO] 9 bytes written for 'stdout'
    proxmox-clone.autogenerated_1: bye
2022/02/22 19:33:54 packer-provisioner-shell plugin: [INFO] 0 bytes written for 'stderr'
2022/02/22 19:33:54 packer-provisioner-shell plugin: [INFO] RPC client: Communicator ended with: 0
2022/02/22 19:33:54 packer-builder-proxmox-clone plugin: [DEBUG] Opening new ssh session
2022/02/22 19:33:54 packer-builder-proxmox-clone plugin: [DEBUG] starting remote command: rm -f /tmp/script_9232.sh
2022/02/22 19:33:54 packer-builder-proxmox-clone plugin: [INFO] RPC endpoint: Communicator ended with: 0
2022/02/22 19:33:54 [INFO] RPC client: Communicator ended with: 0
2022/02/22 19:33:54 [INFO] RPC endpoint: Communicator ended with: 0
2022/02/22 19:33:54 packer-provisioner-shell plugin: [INFO] RPC client: Communicator ended with: 0
2022/02/22 19:33:54 packer-builder-proxmox-clone plugin: [DEBUG] Opening new ssh session
2022/02/22 19:33:54 packer-builder-proxmox-clone plugin: [DEBUG] starting remote command: rm -f
2022/02/22 19:33:54 packer-builder-proxmox-clone plugin: [INFO] RPC endpoint: Communicator ended with: 0
2022/02/22 19:33:54 [INFO] RPC client: Communicator ended with: 0
2022/02/22 19:33:54 [INFO] RPC endpoint: Communicator ended with: 0
2022/02/22 19:33:54 packer-provisioner-shell plugin: [INFO] RPC client: Communicator ended with: 0
2022/02/22 19:33:54 [INFO] (telemetry) ending shell
==> proxmox-clone.autogenerated_1: Trying to remove ephemeral keys from authorized_keys files
2022/02/22 19:33:54 packer-builder-proxmox-clone plugin: [DEBUG] Opening new ssh session
2022/02/22 19:33:54 packer-builder-proxmox-clone plugin: [DEBUG] starting remote command: sed -i.bak '/ packer_62152c64-c1bf-6edb-f245-1312cabd526e$/d' ~/.ssh/authorized_keys; rm ~/.ssh/authorized_keys.bak
2022/02/22 19:33:54 packer-builder-proxmox-clone plugin: [DEBUG] Opening new ssh session
2022/02/22 19:33:54 packer-builder-proxmox-clone plugin: [DEBUG] starting remote command: sudo sed -i.bak '/ packer_62152c64-c1bf-6edb-f245-1312cabd526e$/d' /root/.ssh/authorized_keys; sudo rm /root/.ssh/authorized_keys.bak
==> proxmox-clone.autogenerated_1: Stopping VM
==> proxmox-clone.autogenerated_1: Converting VM to template
2022/02/22 19:33:59 packer-builder-proxmox-clone plugin: template_id: 133
2022/02/22 19:33:59 [INFO] (telemetry) ending proxmox-clone.autogenerated_1
==> Wait completed after 50 seconds 929 milliseconds
==> Builds finished. The artifacts of successful builds are:
Build 'proxmox-clone.autogenerated_1' finished after 50 seconds 929 milliseconds.
2022/02/22 19:33:59 machine readable: proxmox-clone.autogenerated_1,artifact-count []string{"1"}

==> Wait completed after 50 seconds 929 milliseconds

==> Builds finished. The artifacts of successful builds are:
2022/02/22 19:33:59 machine readable: proxmox-clone.autogenerated_1,artifact []string{"0", "builder-id", "proxmox.clone"}
2022/02/22 19:33:59 machine readable: proxmox-clone.autogenerated_1,artifact []string{"0", "id", "133"}
2022/02/22 19:33:59 machine readable: proxmox-clone.autogenerated_1,artifact []string{"0", "string", "A template was created: 133"}
2022/02/22 19:33:59 machine readable: proxmox-clone.autogenerated_1,artifact []string{"0", "files-count", "0"}
2022/02/22 19:33:59 machine readable: proxmox-clone.autogenerated_1,artifact []string{"0", "end"}
--> proxmox-clone.autogenerated_1: A template was created: 133
2022/02/22 19:33:59 [INFO] (telemetry) Finalizing.
2022/02/22 19:34:00 waiting for all plugin processes to complete...
2022/02/22 19:34:00 /home/linuxbrew/.linuxbrew/Cellar/packer/1.7.10/libexec/bin/packer: plugin process exited
2022/02/22 19:34:00 /home/linuxbrew/.linuxbrew/Cellar/packer/1.7.10/libexec/bin/packer: plugin process exited
```

</details>

### LINKED ISSUE #81: Unable to upload ISO with API user
#### Overview of the Issue

When using an API token and either the `cd_files` and/or `cd_content` options in an `additional_iso_files` block like below, or the `iso_url` option, the upload fails with the error `Post "https://<hostname>:8006/api2/json/nodes/<node>/storage/resources/upload": write tcp 192.168.5.71:51734->192.168.1.10:8006: use of closed network connection`. I am able to manually upload an ISO without issue, and I've even tried assigning the API token I used full `Administrator`, but it continues to fail with this error. I can also upload an ISO using a username and password rather than an API token.

```hcl
additional_iso_files {
    device   = "ide3"

    # Generate image with unattended installer
    cd_files = [
        "./autounattend.xml"
    ]
    cd_label = "autounattend"

    iso_storage_pool = "resources"
    unmount          = true
}
```

```hcl
    iso_url          = "./autounattend.iso"
    iso_checksum     = "sha1:539d87c4b66ea7bc9770f103e6e7952a97b56008"
    iso_storage_pool = "resources"
```

I'm also not positive that this report should be here and not `proxmox-api-go`, not sure where the problem lies.

#### Reproduction Steps

Steps to reproduce this issue

1. Use either the `cd_files` block above or create a new `.iso` file and specify it with `iso_url`
2. Create a new API token and give it enough permissions to work with Packer (I even tested with `Administrator`)
3. Run Packer
4. The ISO will immediately fail to upload with the Packer error  `use of closed network connection`
5. Check the `pveproxy` access logs on your Proxmox server, they will show a 403 error and that the request was unauthenticated

### Plugin and Packer version

From `packer version`:

```
Packer v1.8.0
/home/<username>/.config/packer/plugins/github.com/hashicorp/proxmox/packer-plugin-proxmox_v1.0.5_x5.0_linux_amd64
```

### Simplified Packer Buildfile

[Buildfile](https://gist.github.com/dannytech/606a241149716daf22b3aa7586d18f2c)

### Operating system and Environment details

OS, Architecture, and any other information you can provide about the
environment:

- Packer is running in LXC under Proxmox (`uname -a`: `Linux <removed> 5.13.19-6-pve #1 SMP PVE 5.13.19-14 (Thu, 10 Mar 2022 16:24:52 +0100) x86_64 x86_64 x86_64 GNU/Linux`)
- The Packer client and Proxmox server are on different subnets, I had to add a firewall allow so that it could be accessed on `:8006`
- I also proxy the Proxmox interface using Nginx, but the same issue occurred when I tried with that
- The API token uses privilege separation

### Log Fragments and crash.log files

Relevant log from `/var/log/pveproxy/access.log`:

```
::ffff:192.168.5.71 - - [12/04/2022:16:01:19 -0700] "POST /api2/json/nodes/<node>/storage/resources/upload HTTP/1.1" 403 -
```

This error seems to suggest that Packer isn't sending the API credentials, since authenticated requests normally start with `::ffff:127.0.0.1 - root@pam`. So the PVE API would simply be rejecting and closing the POST request, while at the same time Packer tries to upload the file, resulting in the `use of closed network connection`.

### PR DESCRIPTION
**Pull request by sebastian-de**
In preparation of Cloud-Init related additions I'd like to make, I have to update to the latest proxmox-api-go, which includes a fix I committed there: https://github.com/Telmate/proxmox-api-go/pull/216

Since updating the api module requires work already done in #90 I rebased this PR and added a final commit that bumps proxmox-api-go to the latest version. I hope this approach is OK and @alex-sector doesn't mind me taking over.

The updated proxmox-api-go closes #66 
Added support for HTTP heades might also fix #81

### CODE REVIEW

**Comment by sebastian-de** [comment]
I currently can't figure out why `go test -v ./... -run TestLogin` fails, I'm not familiar enough with `net/http/httptest`. Login and Token-auth on a real PVE work though.
I suppose this is because `proxmox-api-go` added support for HTTP headers ( https://github.com/Telmate/proxmox-api-go/pull/202 ).

**Comment by sebastian-de** [comment]
OK, so the error seems to be `http: named cookie not present` upon `req.Cookie("PVEAuthCookie")` (https://github.com/hashicorp/packer-plugin-proxmox/blob/d4a1864f3f2bb68fe0e0bf73f2de7e2650712151/builder/proxmox/common/client_test.go#L67)
@mraerino do you have an idea why?

**Comment by sebastian-de** [comment]
I figued it out myself, the condition now needs to be `req.Header["Authorization"][0] != "PVEAuthCookie=dummy-ticket"`.

**Comment by sebastian-de** [comment]
I also did a rebase of #93 on this PR: https://github.com/sebastian-de/packer-plugin-proxmox/tree/replace
If desired, I can open another PR, should this one get merged.

**Comment by mabeett** [comment]
Hello @sebastian-de,
 during this week [my PR](https://github.com/Telmate/proxmox-api-go/pull/221) which solves #66 was merged.
In case of updating the references to the PR we might solve the referenced issue =)

**Comment by sebastian-de** [comment]
Hi @mabeett I bumped proxmox-api-go to include your fix.

**Comment by sebastian-de** [comment]
Hi @nywilken or @lbajolet-hashicorp (I don't know who else from the Packer-team is currently active here) is there interest from your side to get this merged?
I'd like to make other contributions to this plugin, which are all based on this PR, but I don't know how to proceed.
Should I open PRs which inlude this one, or should I wait until it's merged?

I can totally understand that your resources are limited, but I'm very grateful if anyone can find time for a short feedback.

**Review by lbajolet-hashicorp — CHANGES_REQUESTED** [review]
Hi @sebastian-de,

Sorry for letting this one slide, taking a look at the code right now.
It looks good to me, however I'm slightly worried that the change is breaking, I assume the current code cannot be used with the updated library since there may be some missing information, but I'd be cautious when changing the type of an already existing attribute.

To avoid breaking existing templates, may I suggest adding the new structure as a new attribute in the configs? Something like `efi_config`, and maybe if the idea is to supplant the existing `efidisk` attribute, add a note to the documentation stating that it is deprecated and that the new method is preferred.

That is, unless the current attribute cannot be used with the new structure. If this is the case, we can probably proceed with the changes you propose.

Other than that, the code looks OK to me, we can consider merging this when the PR is amended.

Thanks for the contribution!

**Comment by sebastian-de** [comment]
> To avoid breaking existing templates, may I suggest adding the new structure as a new attribute in the configs? Something like `efi_config`, and maybe if the idea is to supplant the existing `efidisk` attribute, add a note to the documentation stating that it is deprecated and that the new method is preferred.
> 
> That is, unless the current attribute cannot be used with the new structure. If this is the case, we can probably proceed with the changes you propose.

That's a good point. The current attribute cannot be reused - here is the PR for proxmox-api-go that introduced the change: https://github.com/Telmate/proxmox-api-go/pull/179

I'll change the attribute to `efi_config` as you suggested and add the relevant documentation.

File: docs/builders/iso.mdx
Comment (sebastian-de) [reply docs/builders/iso.mdx]: Thanks a lot for your review!
Sorry, I misunderstood your last comment - I removed the `efidisk` section.

Do you create changelogs automatically, or should I add a line to `CHANGELOG.md` with this PR?

File: builder/proxmox/common/config.go
Comment (sebastian-de) [reply builder/proxmox/common/config.go]: I ended up rewriting this part - thanks a lot for your input. Seems that while doing the rebase I didn't spend enough time checking the code code written by the original author.
`PreEnrolledKeys` is now a bool and it's ensured that no device will be created when `efi_config` isn't defined.

**Review by lbajolet-hashicorp — APPROVED** [review]
Hi @sebastian-de,

With the latest reroll, the code looks good to me, thanks for being that responsive, much appreciated!

Given this is a breaking change, I'll wait until someone else from the team approves it as well, as soon as it is approved (unless they have something else to report), we can merge this.

Thanks again!

File: docs/builders/iso.mdx
Comment (lbajolet-hashicorp) [reply docs/builders/iso.mdx]: Changelogs are generated, but we can edit them to add a note if something is breaking for example, so that's probably what will happen when we fold this in, so users are aware of the breaking change.

**Comment by sebastian-de** [comment]
Hey @lbajolet-hashicorp,
that's great to hear!

And thanks to you for the helpful review!

**Comment by sebastian-de** [comment]
I had yet another look at this and found a way to implement the change without breaking existing templates.

The problem was that the old attribute wasn't properly documented anywhere. So I actually searched for templates in the wild that use `efidisk` and found that one, for example: https://github.com/dandyrow/packer-template-proxmox-archlinux/blob/4211ea0b1f96b4db35bd47d45a8253346ce1bfc8/archlinux.source.pkr.hcl#L54

With `efidisk` being simply `efi_storage_pool` it's easy to generate a full `efi_config`, since `efi_storage_pool` is its only mandatory option anyway.

While testing I found that the EFI disk doesn't get created reliably during initial VM creation, so I added an additional step in `stepStartVM`.

Sorry for changing this PR yet again, but avoiding a breaking change makes this worth it, think.

Tested on a PVE 7.3

File: builder/proxmox/common/config.go
Comment (lbajolet-hashicorp) [inline builder/proxmox/common/config.go:357]: Nit regarding phrasing here: 
```suggestion
			errs = packersdk.MultiErrorAppend(errs, fmt.Errorf("both efi_config and efidisk cannot be set at the same time, consider defining only efi_config as efidisk is deprecated"))
```

**Review by lbajolet-hashicorp — APPROVED** [review]
Hi @sebastian-de,

No need to apologise, it's great that you found a way to not break existing configs :)

I just left a last nit regarding the phrasing for the error when both efi_config and efidisk are set at the same time, other than that the code looks good to me.

When the nit is fixed, we can merge this, no need for two approvals since we're not breaking the existing configs now.

Thanks for your perseverance, hopefully we can merge this soon!

File: builder/proxmox/common/config.go
Comment (sebastian-de) [reply builder/proxmox/common/config.go]: Done, thanks!

**Comment by lbajolet-hashicorp** [comment]
Perfect, thanks for the speedy reroll, merging this!

**Pull request merged**
